File: desafio1/fsharp/ToqueFama.fs
open System

let shuffle = List.sortBy(fun _ -> System.Guid.NewGuid())

let validar tam accion =
    let cars = [for c in accion -> c]
    if List.exists (fun c -> not (Char.IsDigit c)) cars then List.empty<int>
    else 
        let org = List.map (fun c -> (int c) - (int '0')) cars
        let num = List.distinct org
        if (List.length org <> List.length num) || (List.length num <> tam) then List.empty<int>
        else num

let comparar num sec = 
    let rec tyf ns xs ys =
        match (ns, xs, ys) with
        | ([], _, _) -> (0,0)
        | (n::ns, x::xs, ys) -> 
            let (t,f) = tyf ns xs ys
            if n = x then (t,f+1) else (if List.exists (fun x -> x = n) ys then (t+1,f) else (t,f))
    tyf num sec sec 

let rec jugar tam sec i =
    printfn "Ingresa una secuencia de %A dígitos distintos (o escribe salir):" tam
    let accion = Console.ReadLine()
    if accion = "salir" then ()
    else 
        let num = validar tam accion
        if List.isEmpty num then 
            printfn "error!\n"
            jugar tam sec (i+1)
        else 
            printfn "ingresaste: %A" num
            let (toques, famas) = comparar num sec
            printfn "resultado: %A Toques, %A Famas\n" toques famas
            if tam = famas then
                printfn "Ganaste! Acertaste al intento %A! La secuencia era %A." i sec
            else
                jugar tam sec (i+1)

let mostrarReglas tam =
    printfn "Bienvenido a Toque y Fama."
    printfn "=========================="
    printfn "En este juego debes tratar de adivinar una secuencia de %d dígitos generada por el programa." tam
    printfn "Para esto ingresas %d dígitos distintos con el fin de adivinar la secuencia."  tam
    printfn "Si has adivinado correctamente la posición de un dígito se produce una Fama."
    printfn "Si has adivinado uno de los dígitos de la secuencia, pero en una posición distinta se trata de un Toque.\n" 
    printfn "Ejemplo: Si la secuencia es secuencia: [8, 0, 6, 1, 3] e ingresas 40863, entonces en pantalla aparecerá:"
    printfn "tu ingresaste [4, 0, 8, 6, 3]" 
    printfn "resultado: 2 Toques 2 Famas\n"


[<EntryPoint>]
let main (argv :string[]) = 
    let tam = 5
    let sec = [ 0 .. 9 ]  |> shuffle |> Seq.take tam |> List.ofSeq
    mostrarReglas tam 
    jugar tam sec 0
    0

File: desafio1/scala/ToqueFama.scala
object ToqueFama {

    def main(args: Array[String]) : Unit = {
        val tam = 5
        val sec = util.Random.shuffle((0 to 9).toList).take(tam).toArray
        println (sec.mkString(", "))
        mostrarReglas(tam)
        var salir = false
        var intentos = 0
        while (!salir) {
            intentos += 1
            println(s"Ingresa una secuencia de ${tam} dígitos distintos (o escribe salir):")
            val accion = scala.io.StdIn.readLine()
            if (accion == null || accion == "salir") {
                println("\ngracias por jugar, adios.")
                salir = true
            }
            else validar(tam, accion) match {
                    case None => println("error!\n")
                    case Some(num) =>
                        println(s"ingresaste: [${num.mkString(", ")}]")
                        val (toques, famas) = comparar(num, sec)
                        println(s"resultado: ${toques} Toques, ${famas} Famas")
                        if (famas == tam) {
                            println(s"Ganaste! Acertaste al intento ${intentos}! La secuencia era [${sec.mkString(", ")}]")
                            salir = true
                        }
                }
        }
    }

    def comparar(num:Array[Int], sec:Array[Int]) : (Int,Int) = {
        val tf = for (n,i) <- num.zipWithIndex; (m,j) <- sec.zipWithIndex if n == m yield if i == j then 'F' else 'T'
        (tf.filter(_ == 'T').length, tf.filter(_ == 'F').length)
    }

    def validar(tam:Int, accion:String) : Option[Array[Int]] = 
        if (accion.exists(!_.isDigit))
            None
        else {
            val num = accion.map(_.toInt - '0'.toInt).distinct
            if (num.length == tam && num.length == accion.length)
                Some(num.toArray)	
            else 
                None	
        }

    def mostrarReglas(tam:Int) : Unit = 
        println(s"""|Bienvenido a Toque y Fama.
                    |==========================\n
                    |En este juego debes tratar de adivinar una secuencia de ${tam} dígitos generadas por el programa.
                    |Para esto ingresas ${tam} dígitos distintos con el fin de adivinar la secuencia.
                    |Si has adivinado correctamente la posición de un dígito se produce una Fama.
                    |Si has adivinado uno de los dígitos de la secuencia, pero en una posición distinta se trata de un Toque.\n
                    |Ejemplo: Si la secuencia es secuencia: [8, 0, 6, 1, 3] e ingresas 40863, entonces en pantalla aparecerá:
                    |tu ingresaste [4, 0, 8, 6, 3]
                    |resultado: 2 Toques 2 Famas\n""".stripMargin)
    
}